editor/code_editor.cpp

```diff
@@ -1608,6 +1618,94 @@ void CodeTextEditor::duplicate_selection() {
+		// The text that will be inserted. All lines in one string
+		String insert_text;
+
+		// The new line position of the caret after the operation
+		int new_caret_line = text_editor->get_caret_line(caret_index);
+		// The new column position of the caret after the operation
+		int new_caret_column = text_editor->get_caret_column(caret_index);
+		// The caret positions of the selection. Stays -1 if there is no selection
+		int select_from_line = -1;
+		int select_to_line = -1;
+		int select_from_column = -1;
+		int select_to_column = -1;
+		// Number of lines of the selection
+		int select_num_lines = -1;
+
+		if (text_editor->has_selection(caret_index)) {
+			select_from_line = text_editor->get_selection_from_line(caret_index);
+			select_to_line = text_editor->get_selection_to_line(caret_index);
+			select_from_column = text_editor->get_selection_from_column(caret_index);
+			select_to_column = text_editor->get_selection_to_column(caret_index);
+			select_num_lines = select_to_line - select_from_line + 1;
+
+			for (int i = select_from_line; i <= select_to_line; i++) {
+				if (p_below) {
+					insert_text += "\n" + text_editor->get_line(i);
+				} else {
+					insert_text += text_editor->get_line(i) + "\n";
+				}
+			}
+			if (p_below) {
+				new_caret_line = select_to_line;
+			} else {
+				new_caret_line = select_from_line;
+				// When above, the new text needs to be inserted at the first line of the selection
+				text_editor->set_caret_line(select_from_line, true, true, 0, caret_index);
+			}
+
+			new_caret_line += select_num_lines - !p_below;
+		} else {
+			if (p_below) {
+				insert_text = "\n" + text_editor->get_line(new_caret_line);
+				new_caret_line++;
+			} else {
+				insert_text = text_editor->get_line(new_caret_line) + "\n";
+			}
+		}
+
+		if (p_below) {
+			// The text will be inserted at the end of the current line
+			text_editor->set_caret_column(text_editor->get_line(text_editor->get_caret_line(caret_index)).length(), true, caret_index);
+		} else {
+			// The text will be inserted at the beginning of the current line
+			text_editor->set_caret_column(0, true, caret_index);
+		}
+
+		if (select_from_line != -1) {
+			for (int i = select_from_line; i <= select_to_line; i++) {
+				text_editor->unfold_line(i);
+			}
+		} else {
+			text_editor->unfold_line(text_editor->get_caret_line(caret_index));
+		}
+		text_editor->deselect(caret_index);
+
+		text_editor->insert_text_at_caret(insert_text, caret_index);
+		text_editor->set_caret_line(new_caret_line, true, true, 0, caret_index);
+		text_editor->set_caret_column(new_caret_column, true, caret_index);
+
+		if (select_from_line != -1) {
+			if (p_below) {
+				// Advance the selection by the number of duplicated lines
+				select_from_line += select_num_lines;
+				select_to_line += select_num_lines;
+			}
+			text_editor->select(select_from_line, select_from_column, select_to_line, select_to_column, caret_index);
+		}
+	}
```

Comment:
Likewise, just put this stuff in the `has_selection()` if/else.

Reply:
I unfortunately can not put this before inserting the text, because I need to deselect when inserting text, because the inserted text replaces the selection.

Reply:
If I could use `_insert_text`, which is private, I could put this before the text insertion, because it wouldn't affect the selection.